### PR TITLE
Auth: Use sessionStorage instead of cookie for automatic redirection

### LINKED ISSUE #92013: Auth: Change how redirection works after login (replace redirect_to cookie)
After a successful login Grafana checks whether the `redirect_to` cookie is included in the request. If it's included then it validates the `redirect_to` url, redirects the user and deletes the cookie. 

## Problem
If Grafana is opened on multiple tabs (with different dashboards) then it can happen that concurrent requests are overwriting/removing the `redirect_to` cookie (after a successful login) and user gets redirected to the root page of Grafana, not to the previously opened dashboard.

## Planned solution
Store the currently opened page URL in the [sessionStorage](https://developer.mozilla.org/en-US/docs/Web/API/Window/sessionStorage)(each tab has its own `sessionStorage`) and _after successful login_ check if the `sessionStorage` contains an url (on the frontend) with a given key and redirect the user to the page.

### PR DESCRIPTION
**Pull request by mgyongyosi**
**What is this feature?**
This PR changes how the redirection happens after a user gets logged out by Grafana. Previously Grafana used the `redirect_to` cookie, but cookies shared by the origin and it doesn't work properly when users have multiple tabs open with f.e. dashboards.

**Why do we need this feature?**

[Add a description of the problem the feature is trying to solve.]

**Who is this feature for?**

[Add information on what kind of user the feature is for.]

**Which issue(s) does this PR fix?**:

Fixes #92013

**Special notes for your reviewer:**

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/contribute/release-notes/#how-to-determine-if-content-belongs-in-whats-new), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/contribute/release-notes/) doc.

### CODE REVIEW

**Comment by joshhunt** [comment]
We need to make sure this works properly when Grafana is being served from a sub path. Both when `serve_from_sub_path` is used to make Grafana's HTTP server serve from a sub path, but also when a reverse proxy is used to serve grafana from a sub path.

File: public/app/app.ts [hunk 1]
```diff
@@ -380,4 +381,49 @@ function reportMetricPerformanceMark(metricName: string, prefix = '', suffix = '
...
+  if (path.startsWith('//')) {
+    return false;
+  }
+
+  if (config.appSubUrl && !path.startsWith(config.appSubUrl) + '/') {
+    return false;
+  }
+
+  return true;
+}
```
Comment (joshhunt) [inline public/app/app.ts:427]: Hmm. Perhaps we should always enforce that we redirect to a path of the origin?

Something like

```js
const redirectParam = sessionStorage.getItem(...);
const redirectTo = window.location.origin + redirectParam;
window.location.replace(redirectTo);
```

?

File: public/app/app.ts [hunk 2]
```diff
@@ -380,4 +381,49 @@ function reportMetricPerformanceMark(metricName: string, prefix = '', suffix = '
   }
 }
 
+function handleRedirectTo(): void {
```
Comment (joshhunt) [inline public/app/app.ts:384]: Why the change to read this on app startup, rather than just issue the 'redirect' to the final destination immediately after login?

Comment (mgyongyosi) [reply public/app/app.ts:384]: >issue the 'redirect' to the final destination immediately after login?

Could you please elaborate?

For more context: there are cases when the redirect url is lost during the requests (OAuth and SAML flows), and previously those backend handlers handled the redirect after login, but with this change (removing the `redirect_to` cookie and using sessionStorage instead) the backend doesn't know where to redirect the user after OAuth/SAML login. (Although we could potentially send the `state` for both SAML and OAuth I decided to implement a solution without changing the `state` params of those)

Comment (joshhunt) [reply public/app/app.ts:384]: Ahhh, I understand. I was not considering oauth flows where the user gets logged in off of grafana, and is redirected back to us.

File: pkg/api/utils.go [hunk 3]
```diff
@@ -7,26 +7,12 @@ import (
...
 
 	"github.com/grafana/grafana/pkg/api/response"
-	"github.com/grafana/grafana/pkg/middleware/cookies"
 	contextmodel "github.com/grafana/grafana/pkg/services/contexthandler/model"
 	"github.com/grafana/grafana/pkg/services/login"
 	"github.com/grafana/grafana/pkg/services/user"
 	"go.opentelemetry.io/otel/trace"
 )
 
-func (hs *HTTPServer) GetRedirectURL(c *contextmodel.ReqContext) string {
```
Comment (linoman) [inline pkg/api/utils.go:17]: This function is used in Grafana Enterprise. [1](https://github.com/grafana/grafana-enterprise/blob/d9f6e949c12978d09b16a64fef89cb904db4e57e/src/pkg/extensions/saml/api.go#L132) [2](https://github.com/grafana/grafana-enterprise/blob/d9f6e949c12978d09b16a64fef89cb904db4e57e/src/pkg/extensions/saml/api.go#L257)

I couldn't find any open branches in Grafana Enterprise for this.

Comment (linoman) [reply pkg/api/utils.go:17]: Found it! https://github.com/grafana/grafana-enterprise/pull/7141

**Comment by mgyongyosi** [comment]
> We need to make sure this works properly when Grafana is being served from a sub path. Both when `serve_from_sub_path` is used to make Grafana's HTTP server serve from a sub path, but also when a reverse proxy is used to serve grafana from a sub path.

I tested these scenarios and the redirection seemed to be worked correctly.

**Review by linoman — APPROVED** [review]
I've tested this feature with different scenarios and browsers. It looks great and handles things correctly when other tabs are open.

Great job!

**Comment by mgyongyosi** [comment]
/deploy-to-hg --enterprise-ref mgyongyosi/redirect-session-storage

**Review by joshhunt — APPROVED** [review]
Looks _great_ to me! Thanks for your patience 😊

**Comment by mgyongyosi** [comment]
/deploy-to-hg --enterprise-ref mgyongyosi/redirect-session-storage

File: pkg/services/accesscontrol/middleware.go [hunk 4]
```diff
@@ -162,6 +184,21 @@ func writeRedirectCookie(c *contextmodel.ReqContext) {
...
+
+	if redirectTo == "/" {
+		return ""
+	}
+
+	// remove any forceLogin=true params
+	redirectTo = removeForceLoginParams(redirectTo)
+	return "?redirectTo=" + url.QueryEscape(redirectTo)
+}
+
```
Comment (kalleep) [inline pkg/services/accesscontrol/middleware.go:201]: A lot of this is duplicated between the access control and auth middleware, we should probably extract common functionality somewhere

File: pkg/services/contexthandler/model/model.go [hunk 5]
```diff
@@ -35,6 +35,10 @@ type ReqContext struct {
 
 	PerfmonTimer   prometheus.Summary
 	LookupTokenErr error
+
+	// FIXME: Remove this temporary flag after the rollout of FlagUseSessionStorageRedirect feature flag
+	// UseSessionStorageRedirect is introduced to simplify the rollout of the new redirect logic
```
Comment (kalleep) [inline pkg/services/contexthandler/model/model.go:40]: Lets create a issue for removing this and link it here in the comment

**Review by kalleep — APPROVED** [review]
LGTM with one nit

File: pkg/services/contexthandler/model/model.go (hunk 5 above)
Comment (mgyongyosi) [reply pkg/services/contexthandler/model/model.go]: Done
